File: DSA-lab/modules/01-komplexitaetsanalyse-big-o/projects/03-final/audit_pipeline.py
"""Baseline and optimized customer aggregation pipelines."""

from __future__ import annotations

import csv
from dataclasses import dataclass, field
from pathlib import Path
from typing import Iterable, Mapping


@dataclass(frozen=True)
class Event:
    """One immutable event loaded from the source CSV."""

    event_id: str
    customer_id: str
    category: str
    amount_cents: int
    discount_cents: int
    status: str
    event_timestamp: str

    @classmethod
    def from_csv_row(cls, row: Mapping[str, str]) -> "Event":
        """Build and validate an event from a CSV mapping."""
        amount_cents = int(row["amount_cents"])
        discount_cents = int(row["discount_cents"])
        if amount_cents < 0:
            raise ValueError("amount_cents must not be negative")
        if not 0 <= discount_cents <= amount_cents:
            raise ValueError("discount_cents must be between zero and amount")

        return cls(
            event_id=row["event_id"],
            customer_id=row["customer_id"],
            category=row["category"],
            amount_cents=amount_cents,
            discount_cents=discount_cents,
            status=row["status"],
            event_timestamp=row["event_timestamp"],
        )

    @property
    def net_cents(self) -> int:
        """Return revenue after discount."""
        return self.amount_cents - self.discount_cents


@dataclass(frozen=True)
class CustomerSummary:
    """Aggregated metrics for one customer."""

    customer_id: str
    completed_events: int
    net_revenue_cents: int
    average_net_cents: int
    unique_categories: int
    latest_event_timestamp: str


@dataclass
class _Accumulator:
    """Mutable aggregation state used by the optimized pipeline."""

    completed_events: int = 0
    net_revenue_cents: int = 0
    categories: set[str] = field(default_factory=set)
    latest_event_timestamp: str = ""

    def update(self, event: Event) -> None:
        """Consume one completed event."""
        self.completed_events += 1
        self.net_revenue_cents += event.net_cents
        self.categories.add(event.category)
        if event.event_timestamp > self.latest_event_timestamp:
            self.latest_event_timestamp = event.event_timestamp

    def to_summary(self, customer_id: str) -> CustomerSummary:
        """Freeze accumulated values into an output row."""
        return CustomerSummary(
            customer_id=customer_id,
            completed_events=self.completed_events,
            net_revenue_cents=self.net_revenue_cents,
            average_net_cents=(
                self.net_revenue_cents // self.completed_events
            ),
            unique_categories=len(self.categories),
            latest_event_timestamp=self.latest_event_timestamp,
        )


def load_events(source: Path) -> list[Event]:
    """Load all events from a CSV file."""
    with source.open(newline="", encoding="utf-8") as csv_file:
        reader = csv.DictReader(csv_file)
        return [Event.from_csv_row(row) for row in reader]


def inefficient_pipeline(events: Iterable[Event]) -> list[CustomerSummary]:
    """Aggregate through list membership checks and repeated full scans."""
    event_list = list(events)
    customer_ids: list[str] = []

    for event in event_list:
        if (
            event.status == "completed"
            and event.customer_id not in customer_ids
        ):
            customer_ids.append(event.customer_id)

    summaries = []
    for customer_id in customer_ids:
        customer_events = [
            event
            for event in event_list
            if (
                event.status == "completed"
                and event.customer_id == customer_id
            )
        ]
        net_revenue_cents = sum(event.net_cents for event in customer_events)
        categories = {event.category for event in customer_events}
        latest_timestamp = max(
            event.event_timestamp for event in customer_events
        )

        summaries.append(
            CustomerSummary(
                customer_id=customer_id,
                completed_events=len(customer_events),
                net_revenue_cents=net_revenue_cents,
                average_net_cents=(
                    net_revenue_cents // len(customer_events)
                ),
                unique_categories=len(categories),
                latest_event_timestamp=latest_timestamp,
            )
        )

    return sorted(summaries, key=lambda summary: summary.customer_id)


def optimized_pipeline(events: Iterable[Event]) -> list[CustomerSummary]:
    """Aggregate all completed events in one pass with hash lookups."""
    accumulators: dict[str, _Accumulator] = {}

    for event in events:
        if event.status != "completed":
            continue

        accumulator = accumulators.get(event.customer_id)
        if accumulator is None:
            accumulator = _Accumulator()
            accumulators[event.customer_id] = accumulator
        accumulator.update(event)

    return [
        accumulators[customer_id].to_summary(customer_id)
        for customer_id in sorted(accumulators)
    ]


def write_summaries(
    summaries: Iterable[CustomerSummary],
    destination: Path,
) -> None:
    """Write customer summaries to CSV."""
    destination.parent.mkdir(parents=True, exist_ok=True)
    fieldnames = (
        "customer_id",
        "completed_events",
        "net_revenue_cents",
        "average_net_cents",
        "unique_categories",
        "latest_event_timestamp",
    )

    with destination.open("w", newline="", encoding="utf-8") as csv_file:
        writer = csv.DictWriter(csv_file, fieldnames=fieldnames)
        writer.writeheader()
        for summary in summaries:
            writer.writerow(
                {
                    fieldname: getattr(summary, fieldname)
                    for fieldname in fieldnames
                }
            )

File: DSA-lab-java/modules/01-komplexitaetsanalyse-big-o/projects/03-final/src/main/java/lab/dsa/module01/finalproject/AuditPipeline.java
package lab.dsa.module01.finalproject;

import java.io.BufferedWriter;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

/** Baseline and optimized customer aggregation pipelines. */
public final class AuditPipeline {
    private AuditPipeline() {
    }

    /** One immutable event loaded from the source CSV. */
    public record Event(String eventId, String customerId, String category,
                        int amountCents, int discountCents, String status,
                        String eventTimestamp) {
        public Event {
            if (amountCents < 0) {
                throw new IllegalArgumentException("Amount must not be negative");
            }
            if (discountCents < 0 || discountCents > amountCents) {
                throw new IllegalArgumentException("Discount must be between zero and amount");
            }
        }

        public int netCents() {
            return amountCents - discountCents;
        }
    }

    /** Aggregated metrics for one customer. */
    public record CustomerSummary(String customerId, int completedEvents,
                                  long netRevenueCents, long averageNetCents,
                                  int uniqueCategories, String latestEventTimestamp) {
    }

    private static final class Accumulator {
        private int completedEvents;
        private long netRevenueCents;
        private final Set<String> categories = new HashSet<>();
        private String latestEventTimestamp = "";

        private void update(Event event) {
            completedEvents++;
            netRevenueCents += event.netCents();
            categories.add(event.category());
            if (event.eventTimestamp().compareTo(latestEventTimestamp) > 0) {
                latestEventTimestamp = event.eventTimestamp();
            }
        }

        private CustomerSummary toSummary(String customerId) {
            return new CustomerSummary(customerId, completedEvents, netRevenueCents,
                    netRevenueCents / completedEvents, categories.size(), latestEventTimestamp);
        }
    }

    public static List<Event> loadEvents(Path source) throws IOException {
        List<String> lines = Files.readAllLines(source, StandardCharsets.UTF_8);
        if (lines.isEmpty()) {
            return List.of();
        }
        String expectedHeader = "event_id,customer_id,category,amount_cents,discount_cents,status,event_timestamp";
        if (!lines.getFirst().equals(expectedHeader)) {
            throw new IllegalArgumentException("Unexpected event CSV header");
        }
        List<Event> events = new ArrayList<>(lines.size() - 1);
        for (int lineNumber = 1; lineNumber < lines.size(); lineNumber++) {
            String[] fields = lines.get(lineNumber).split(",", -1);
            if (fields.length != 7) {
                throw new IllegalArgumentException("Invalid event row at line " + (lineNumber + 1));
            }
            events.add(new Event(fields[0], fields[1], fields[2], Integer.parseInt(fields[3]),
                    Integer.parseInt(fields[4]), fields[5], fields[6]));
        }
        return events;
    }

    /** Aggregates through list membership checks and repeated full scans. */
    public static List<CustomerSummary> inefficientPipeline(List<Event> events) {
        List<Event> eventList = new ArrayList<>(events);
        List<String> customerIds = new ArrayList<>();
        for (Event event : eventList) {
            if (event.status().equals("completed") && !customerIds.contains(event.customerId())) {
                customerIds.add(event.customerId());
            }
        }

        List<CustomerSummary> summaries = new ArrayList<>();
        for (String customerId : customerIds) {
            List<Event> customerEvents = new ArrayList<>();
            for (Event event : eventList) {
                if (event.status().equals("completed") && event.customerId().equals(customerId)) {
                    customerEvents.add(event);
                }
            }
            long netRevenueCents = 0;
            Set<String> categories = new HashSet<>();
            String latestTimestamp = "";
            for (Event event : customerEvents) {
                netRevenueCents += event.netCents();
                categories.add(event.category());
                if (event.eventTimestamp().compareTo(latestTimestamp) > 0) {
                    latestTimestamp = event.eventTimestamp();
                }
            }
            summaries.add(new CustomerSummary(customerId, customerEvents.size(), netRevenueCents,
                    netRevenueCents / customerEvents.size(), categories.size(), latestTimestamp));
        }
        summaries.sort(Comparator.comparing(CustomerSummary::customerId));
        return summaries;
    }

    /** Aggregates all completed events in one pass with hash lookups. */
    public static List<CustomerSummary> optimizedPipeline(List<Event> events) {
        Map<String, Accumulator> accumulators = new HashMap<>();
        for (Event event : events) {
            if (!event.status().equals("completed")) {
                continue;
            }
            accumulators.computeIfAbsent(event.customerId(), ignored -> new Accumulator()).update(event);
        }
        return accumulators.keySet().stream().sorted()
                .map(customerId -> accumulators.get(customerId).toSummary(customerId)).toList();
    }

    public static void writeSummaries(List<CustomerSummary> summaries, Path destination) throws IOException {
        if (destination.getParent() != null) {
            Files.createDirectories(destination.getParent());
        }
        try (BufferedWriter writer = Files.newBufferedWriter(destination, StandardCharsets.UTF_8)) {
            writer.write("customer_id,completed_events,net_revenue_cents,average_net_cents,unique_categories,latest_event_timestamp");
            writer.newLine();
            for (CustomerSummary summary : summaries) {
                writer.write(String.join(",", summary.customerId(),
                        Integer.toString(summary.completedEvents()), Long.toString(summary.netRevenueCents()),
                        Long.toString(summary.averageNetCents()), Integer.toString(summary.uniqueCategories()),
                        summary.latestEventTimestamp()));
                writer.newLine();
            }
        }
    }
}
